QueryLambdaClient closes System.in

### Description / Background

The script `scripts/utility/lambdaQuery.sh` throws an exception because `System.in` has been closed.

### Steps to reproduce

1. Deploy a Sleeper instance with the query stack
2. Run `scripts/utility/lambdaQuery.sh <instance-id>`
3. Choose query type
4. See error

### Expected behaviour

The script should run a query against the Sleeper instance.

### Technical Notes / Implementation Details

It looks like this was caused by a change to QueryLambdaClient in the following PR:
- https://github.com/gchq/sleeper/pull/7365

System.in is being passed into a Scanner in a try-with-resources, which seems to be closing System.in.

### Screenshots/Logs

Error from script:
```
Exact (e) or range (r) query? Exception in thread "main" java.io.IOError: java.io.IOException: Stream Closed
        at java.base/java.io.Console.readLine(Console.java:258)
        at sleeper.clients.util.console.ConsoleInput.promptLine(ConsoleInput.java:45)
        at sleeper.clients.query.QueryCommandLineClient.runQueries(QueryCommandLineClient.java:112)
        at sleeper.clients.query.QueryLambdaClient.runQueries(QueryLambdaClient.java:125)
        at sleeper.clients.query.QueryCommandLineClient.run(QueryCommandLineClient.java:91)
        at sleeper.clients.query.QueryLambdaClient.main(QueryLambdaClient.java:147)
Caused by: java.io.IOException: Stream Closed
        at java.base/java.io.FileInputStream.readBytes(Native Method)
        at java.base/java.io.FileInputStream.read(FileInputStream.java:276)
        at java.base/sun.nio.cs.StreamDecoder.readBytes(StreamDecoder.java:287)
        at java.base/sun.nio.cs.StreamDecoder.implRead(StreamDecoder.java:330)
        at java.base/sun.nio.cs.StreamDecoder.read(StreamDecoder.java:190)
        at java.base/java.io.Console$LineReader.read(Console.java:511)
        at java.base/java.io.Console.readline(Console.java:442)
        at java.base/java.io.Console.readLine(Console.java:254)
        ... 5 more
```

--- a/java/clients/src/main/java/sleeper/clients/api/QuerySender.java
+++ b/java/clients/src/main/java/sleeper/clients/api/QuerySender.java
@@ -1,0 +1,55 @@
+/*
+ * Copyright 2022-2026 Crown Copyright
+ *
+ * Licensed under the Apache License, Version 2.0 (the "License");
+ * you may not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ *
+ *     http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+package sleeper.clients.api;
+
+import software.amazon.awssdk.services.sqs.SqsClient;
+
+import sleeper.core.properties.instance.InstanceProperties;
+import sleeper.core.properties.table.TablePropertiesProvider;
+import sleeper.query.core.model.Query;
+import sleeper.query.core.model.QuerySerDe;
+
+import static sleeper.core.properties.instance.CdkDefinedInstanceProperty.QUERY_QUEUE_URL;
+
+/**
+ * Sends a query to a Sleeper table. The query must be tracked and results retrieved separately.
+ */
+public interface QuerySender {
+
+    /**
+     * Sends a query against a Sleeper table.
+     *
+     * @param query the query
+     */
+    void sendQuery(Query query);
+
+    /**
+     * Creates a sender that will submit queries to an SQS queue.
+     *
+     * @param  instanceProperties      the instance properties
+     * @param  tablePropertiesProvider the table properties provider
+     * @param  sqsClient               the SQS client
+     * @return                         the sender
+     */
+    static QuerySender toSqs(InstanceProperties instanceProperties, TablePropertiesProvider tablePropertiesProvider, SqsClient sqsClient) {
+        QuerySerDe querySerDe = new QuerySerDe(tablePropertiesProvider);
+        return query -> {
+            sqsClient.sendMessage(request -> request.queueUrl(instanceProperties.get(QUERY_QUEUE_URL))
+                    .messageBody(querySerDe.toJson(query)));
+        };
+    }
+
+}

--- a/java/clients/src/main/java/sleeper/clients/query/QueryCommandLineClient.java
+++ b/java/clients/src/main/java/sleeper/clients/query/QueryCommandLineClient.java
@@ -16,13 +16,9 @@
 package sleeper.clients.query;
 
 import org.apache.commons.codec.binary.Base64;
-import software.amazon.awssdk.services.dynamodb.DynamoDbClient;
-import software.amazon.awssdk.services.s3.S3Client;
 
 import sleeper.clients.util.console.ConsoleInput;
 import sleeper.clients.util.console.ConsoleOutput;
-import sleeper.configuration.properties.S3TableProperties;
-import sleeper.configuration.table.index.DynamoDBTableIndex;
 import sleeper.core.properties.instance.InstanceProperties;
 import sleeper.core.properties.table.TableProperties;
 import sleeper.core.properties.table.TablePropertiesProvider;
@@ -59,11 +55,6 @@ public abstract class QueryCommandLineClient {
     private final Supplier<String> queryIdSupplier;
     protected final ConsoleInput in;
     protected final ConsoleOutput out;
-
-    protected QueryCommandLineClient(S3Client s3Client, DynamoDbClient dynamoClient, InstanceProperties instanceProperties,
-            ConsoleInput in, ConsoleOutput out) {
-        this(instanceProperties, new DynamoDBTableIndex(instanceProperties, dynamoClient), S3TableProperties.createProvider(instanceProperties, s3Client, dynamoClient), in, out);
-    }
 
     protected QueryCommandLineClient(InstanceProperties instanceProperties, TableIndex tableIndex, TablePropertiesProvider tablePropertiesProvider,
             ConsoleInput in, ConsoleOutput out) {

--- a/java/clients/src/main/java/sleeper/clients/query/QueryCommandLineClient.java
+++ b/java/clients/src/main/java/sleeper/clients/query/QueryCommandLineClient.java
@@ -60,10 +60,6 @@ public abstract class QueryCommandLineClient {
     protected ConsoleInput in;
     protected ConsoleOutput out;
 
-    protected QueryCommandLineClient(S3Client s3Client, DynamoDbClient dynamoClient, InstanceProperties instanceProperties) {
-        this(s3Client, dynamoClient, instanceProperties, ConsoleInput.stdIn(), ConsoleOutput.stdOut());
-    }
-
     protected QueryCommandLineClient(S3Client s3Client, DynamoDbClient dynamoClient, InstanceProperties instanceProperties,
             ConsoleInput in, ConsoleOutput out) {
         this(instanceProperties, new DynamoDBTableIndex(instanceProperties, dynamoClient), S3TableProperties.createProvider(instanceProperties, s3Client, dynamoClient), in, out);

--- a/java/clients/src/main/java/sleeper/clients/query/QueryCommandLineClient.java
+++ b/java/clients/src/main/java/sleeper/clients/query/QueryCommandLineClient.java
@@ -57,8 +57,8 @@ public abstract class QueryCommandLineClient {
     private final TablePropertiesProvider tablePropertiesProvider;
     private final InstanceProperties instanceProperties;
     private final Supplier<String> queryIdSupplier;
-    protected ConsoleInput in;
-    protected ConsoleOutput out;
+    protected final ConsoleInput in;
+    protected final ConsoleOutput out;
 
     protected QueryCommandLineClient(S3Client s3Client, DynamoDbClient dynamoClient, InstanceProperties instanceProperties,
             ConsoleInput in, ConsoleOutput out) {

--- a/java/clients/src/main/java/sleeper/clients/query/QueryLambdaClient.java
+++ b/java/clients/src/main/java/sleeper/clients/query/QueryLambdaClient.java
@@ -47,8 +47,6 @@ import static sleeper.core.properties.instance.CdkDefinedInstanceProperty.QUERY_
  * execute the query.
  */
 public class QueryLambdaClient extends QueryCommandLineClient {
-    private final ConsoleOutput out;
-    private final ConsoleInput in;
     private final SqsClient sqsClient;
     private final DynamoDBQueryTracker queryTracker;
     private Map<String, String> resultsPublisherConfig;
@@ -56,9 +54,7 @@ public class QueryLambdaClient extends QueryCommandLineClient {
     private final QuerySerDe querySerDe;
 
     public QueryLambdaClient(InstanceProperties instanceProperties, S3Client s3Client, DynamoDbClient dynamoClient, SqsClient sqsClient, ConsoleOutput out, ConsoleInput in) {
-        super(s3Client, dynamoClient, instanceProperties);
-        this.out = out;
-        this.in = in;
+        super(s3Client, dynamoClient, instanceProperties, in, out);
         this.sqsClient = sqsClient;
         this.queryTracker = new DynamoDBQueryTracker(instanceProperties, dynamoClient);
         this.queryQueueUrl = instanceProperties.get(QUERY_QUEUE_URL);

--- a/java/clients/src/main/java/sleeper/clients/query/QueryLambdaClient.java
+++ b/java/clients/src/main/java/sleeper/clients/query/QueryLambdaClient.java
@@ -20,17 +20,21 @@ import software.amazon.awssdk.services.s3.S3Client;
 import software.amazon.awssdk.services.sqs.SqsClient;
 import software.amazon.awssdk.services.sts.StsClient;
 
+import sleeper.clients.api.QuerySender;
 import sleeper.clients.util.console.ConsoleInput;
 import sleeper.clients.util.console.ConsoleOutput;
 import sleeper.configuration.properties.S3InstanceProperties;
 import sleeper.configuration.properties.S3TableProperties;
+import sleeper.configuration.table.index.DynamoDBTableIndex;
 import sleeper.core.properties.instance.InstanceProperties;
 import sleeper.core.properties.table.TableProperties;
+import sleeper.core.properties.table.TablePropertiesProvider;
+import sleeper.core.table.TableIndex;
 import sleeper.query.core.model.Query;
-import sleeper.query.core.model.QuerySerDe;
 import sleeper.query.core.output.ResultsOutput;
 import sleeper.query.core.tracker.QueryState;
 import sleeper.query.core.tracker.QueryTrackerException;
+import sleeper.query.core.tracker.QueryTrackerStore;
 import sleeper.query.core.tracker.TrackedQuery;
 import sleeper.query.runner.output.SQSResultsOutput;
 import sleeper.query.runner.tracker.DynamoDBQueryTracker;
@@ -38,7 +42,6 @@ import sleeper.query.runner.tracker.DynamoDBQueryTracker;
 import java.util.HashMap;
 import java.util.Map;
 
-import static sleeper.core.properties.instance.CdkDefinedInstanceProperty.QUERY_QUEUE_URL;
 import static sleeper.core.properties.instance.CdkDefinedInstanceProperty.QUERY_RESULTS_BUCKET;
 import static sleeper.core.properties.instance.CdkDefinedInstanceProperty.QUERY_RESULTS_QUEUE_URL;
 
@@ -47,18 +50,16 @@ import static sleeper.core.properties.instance.CdkDefinedInstanceProperty.QUERY_
  * execute the query.
  */
 public class QueryLambdaClient extends QueryCommandLineClient {
-    private final SqsClient sqsClient;
-    private final DynamoDBQueryTracker queryTracker;
+    private final QuerySender querySender;
+    private final QueryTrackerStore queryTracker;
     private Map<String, String> resultsPublisherConfig;
-    private final String queryQueueUrl;
-    private final QuerySerDe querySerDe;
 
-    public QueryLambdaClient(InstanceProperties instanceProperties, S3Client s3Client, DynamoDbClient dynamoClient, SqsClient sqsClient, ConsoleOutput out, ConsoleInput in) {
-        super(s3Client, dynamoClient, instanceProperties, in, out);
-        this.sqsClient = sqsClient;
-        this.queryTracker = new DynamoDBQueryTracker(instanceProperties, dynamoClient);
-        this.queryQueueUrl = instanceProperties.get(QUERY_QUEUE_URL);
-        this.querySerDe = new QuerySerDe(S3TableProperties.createProvider(instanceProperties, s3Client, dynamoClient));
+    public QueryLambdaClient(
+            InstanceProperties instanceProperties, TableIndex tableIndex, TablePropertiesProvider tablePropertiesProvider,
+            QuerySender querySender, QueryTrackerStore queryTracker, ConsoleInput in, ConsoleOutput out) {
+        super(instanceProperties, tableIndex, tablePropertiesProvider, in, out);
+        this.querySender = querySender;
+        this.queryTracker = queryTracker;
     }
 
     @Override
@@ -123,9 +124,7 @@ public class QueryLambdaClient extends QueryCommandLineClient {
     }
 
     public void submitQuery(Query query) {
-        sqsClient.sendMessage(request -> request.queueUrl(queryQueueUrl)
-                .messageBody(querySerDe.toJson(
-                        query.withResultsPublisherConfig(resultsPublisherConfig))));
+        querySender.sendQuery(query.withResultsPublisherConfig(resultsPublisherConfig));
     }
 
     public static void main(String[] args) throws InterruptedException {
@@ -140,7 +139,13 @@ public class QueryLambdaClient extends QueryCommandLineClient {
                 StsClient stsClient = StsClient.create()) {
             String accountName = stsClient.getCallerIdentity().account();
             InstanceProperties instanceProperties = S3InstanceProperties.loadGivenAccountAndInstanceId(s3Client, accountName, instanceId);
-            QueryLambdaClient queryLambdaClient = new QueryLambdaClient(instanceProperties, s3Client, dynamoClient, sqsClient, ConsoleOutput.stdOut(), ConsoleInput.stdIn());
+            TableIndex tableIndex = new DynamoDBTableIndex(instanceProperties, dynamoClient);
+            TablePropertiesProvider tablePropertiesProvider = S3TableProperties.createProvider(instanceProperties, tableIndex, s3Client);
+            QueryLambdaClient queryLambdaClient = new QueryLambdaClient(
+                    instanceProperties, tableIndex, tablePropertiesProvider,
+                    QuerySender.toSqs(instanceProperties, tablePropertiesProvider, sqsClient),
+                    new DynamoDBQueryTracker(instanceProperties, dynamoClient),
+                    ConsoleInput.stdIn(), ConsoleOutput.stdOut());
             queryLambdaClient.run();
         }
     }

--- a/java/clients/src/main/java/sleeper/clients/query/QueryLambdaClient.java
+++ b/java/clients/src/main/java/sleeper/clients/query/QueryLambdaClient.java
@@ -21,6 +21,7 @@ import software.amazon.awssdk.services.sqs.SqsClient;
 import software.amazon.awssdk.services.sts.StsClient;
 
 import sleeper.clients.util.console.ConsoleInput;
+import sleeper.clients.util.console.ConsoleOutput;
 import sleeper.configuration.properties.S3InstanceProperties;
 import sleeper.configuration.properties.S3TableProperties;
 import sleeper.core.properties.instance.InstanceProperties;
@@ -46,6 +47,7 @@ import static sleeper.core.properties.instance.CdkDefinedInstanceProperty.QUERY_
  * execute the query.
  */
 public class QueryLambdaClient extends QueryCommandLineClient {
+    private final ConsoleOutput out;
     private final ConsoleInput in;
     private final SqsClient sqsClient;
     private final DynamoDBQueryTracker queryTracker;
@@ -53,8 +55,9 @@ public class QueryLambdaClient extends QueryCommandLineClient {
     private final String queryQueueUrl;
     private final QuerySerDe querySerDe;
 
-    public QueryLambdaClient(InstanceProperties instanceProperties, S3Client s3Client, DynamoDbClient dynamoClient, SqsClient sqsClient, ConsoleInput in) {
+    public QueryLambdaClient(InstanceProperties instanceProperties, S3Client s3Client, DynamoDbClient dynamoClient, SqsClient sqsClient, ConsoleOutput out, ConsoleInput in) {
         super(s3Client, dynamoClient, instanceProperties);
+        this.out = out;
         this.in = in;
         this.sqsClient = sqsClient;
         this.queryTracker = new DynamoDBQueryTracker(instanceProperties, dynamoClient);
@@ -69,14 +72,14 @@ public class QueryLambdaClient extends QueryCommandLineClient {
 
     @Override
     protected void submitQuery(TableProperties tableProperties, Query query) {
-        System.out.println("Submitting query with id " + query.getQueryId());
+        out.println("Submitting query with id " + query.getQueryId());
         submitQuery(query);
         long sleepTime = 1000L;
         try {
             QueryState state;
             int count = 0;
             while (true) {
-                System.out.println("Polling query tracker");
+                out.println("Polling query tracker");
                 TrackedQuery trackedQuery = queryTracker.getStatus(query.getQueryId());
                 if (trackedQuery != null) {
                     state = trackedQuery.getLastKnownState();
@@ -90,12 +93,12 @@ public class QueryLambdaClient extends QueryCommandLineClient {
                 } else if (count > 10) {
                     sleepTime = 2000L;
                 }
-                System.out.println("Sleeping for " + (sleepTime / 1000) + " seconds");
+                out.println("Sleeping for " + (sleepTime / 1000) + " seconds");
                 Thread.sleep(sleepTime);
             }
-            System.out.println("Finished query processing with final state of: " + state);
+            out.println("Finished query processing with final state of: " + state);
         } catch (QueryTrackerException | InterruptedException e) {
-            System.out.println("Failed to get status");
+            out.println("Failed to get status");
             e.printStackTrace();
         }
     }
@@ -111,10 +114,10 @@ public class QueryLambdaClient extends QueryCommandLineClient {
             if ("s".equalsIgnoreCase(type)) {
                 // Nothing to do - empty resultsPublisherConfig will cause the
                 // results to be published to S3.
-                System.out.println("Results will be published to S3 bucket " + getInstanceProperties().get(QUERY_RESULTS_BUCKET));
+                out.println("Results will be published to S3 bucket " + getInstanceProperties().get(QUERY_RESULTS_BUCKET));
             } else if ("q".equals(type)) {
                 resultsPublisherConfig.put(ResultsOutput.DESTINATION, SQSResultsOutput.SQS);
-                System.out.println("Results will be published to SQS queue " + getInstanceProperties().get(QUERY_RESULTS_QUEUE_URL));
+                out.println("Results will be published to SQS queue " + getInstanceProperties().get(QUERY_RESULTS_QUEUE_URL));
             } else {
                 continue;
             }
@@ -141,7 +144,7 @@ public class QueryLambdaClient extends QueryCommandLineClient {
                 StsClient stsClient = StsClient.create()) {
             String accountName = stsClient.getCallerIdentity().account();
             InstanceProperties instanceProperties = S3InstanceProperties.loadGivenAccountAndInstanceId(s3Client, accountName, instanceId);
-            QueryLambdaClient queryLambdaClient = new QueryLambdaClient(instanceProperties, s3Client, dynamoClient, sqsClient, ConsoleInput.stdIn());
+            QueryLambdaClient queryLambdaClient = new QueryLambdaClient(instanceProperties, s3Client, dynamoClient, sqsClient, ConsoleOutput.stdOut(), ConsoleInput.stdIn());
             queryLambdaClient.run();
         }
     }

--- a/java/clients/src/main/java/sleeper/clients/query/QueryLambdaClient.java
+++ b/java/clients/src/main/java/sleeper/clients/query/QueryLambdaClient.java
@@ -20,6 +20,7 @@ import software.amazon.awssdk.services.s3.S3Client;
 import software.amazon.awssdk.services.sqs.SqsClient;
 import software.amazon.awssdk.services.sts.StsClient;
 
+import sleeper.clients.util.console.ConsoleInput;
 import sleeper.configuration.properties.S3InstanceProperties;
 import sleeper.configuration.properties.S3TableProperties;
 import sleeper.core.properties.instance.InstanceProperties;
@@ -33,10 +34,8 @@ import sleeper.query.core.tracker.TrackedQuery;
 import sleeper.query.runner.output.SQSResultsOutput;
 import sleeper.query.runner.tracker.DynamoDBQueryTracker;
 
-import java.nio.charset.StandardCharsets;
 import java.util.HashMap;
 import java.util.Map;
-import java.util.Scanner;
 
 import static sleeper.core.properties.instance.CdkDefinedInstanceProperty.QUERY_QUEUE_URL;
 import static sleeper.core.properties.instance.CdkDefinedInstanceProperty.QUERY_RESULTS_BUCKET;
@@ -47,14 +46,16 @@ import static sleeper.core.properties.instance.CdkDefinedInstanceProperty.QUERY_
  * execute the query.
  */
 public class QueryLambdaClient extends QueryCommandLineClient {
+    private final ConsoleInput in;
     private final SqsClient sqsClient;
     private final DynamoDBQueryTracker queryTracker;
     private Map<String, String> resultsPublisherConfig;
     private final String queryQueueUrl;
     private final QuerySerDe querySerDe;
 
-    public QueryLambdaClient(S3Client s3Client, DynamoDbClient dynamoClient, SqsClient sqsClient, InstanceProperties instanceProperties) {
+    public QueryLambdaClient(InstanceProperties instanceProperties, S3Client s3Client, DynamoDbClient dynamoClient, SqsClient sqsClient, ConsoleInput in) {
         super(s3Client, dynamoClient, instanceProperties);
+        this.in = in;
         this.sqsClient = sqsClient;
         this.queryTracker = new DynamoDBQueryTracker(instanceProperties, dynamoClient);
         this.queryQueueUrl = instanceProperties.get(QUERY_QUEUE_URL);
@@ -102,25 +103,22 @@ public class QueryLambdaClient extends QueryCommandLineClient {
     @Override
     protected void runQueries(TableProperties tableProperties) throws InterruptedException {
         resultsPublisherConfig = new HashMap<>();
-        try (Scanner scanner = new Scanner(System.in, StandardCharsets.UTF_8)) {
-            while (true) {
-                System.out.println("Send output to S3 results bucket (s) or SQS (q)?");
-                String type = scanner.nextLine();
-                if ("".equals(type)) {
-                    break;
-                }
-                if ("s".equalsIgnoreCase(type)) {
-                    // Nothing to do - empty resultsPublisherConfig will cause the
-                    // results to be published to S3.
-                    System.out.println("Results will be published to S3 bucket " + getInstanceProperties().get(QUERY_RESULTS_BUCKET));
-                } else if ("q".equals(type)) {
-                    resultsPublisherConfig.put(ResultsOutput.DESTINATION, SQSResultsOutput.SQS);
-                    System.out.println("Results will be published to SQS queue " + getInstanceProperties().get(QUERY_RESULTS_QUEUE_URL));
-                } else {
-                    continue;
-                }
+        while (true) {
+            String type = in.promptLine("Send output to S3 results bucket (s) or SQS (q)?");
+            if ("".equals(type)) {
                 break;
             }
+            if ("s".equalsIgnoreCase(type)) {
+                // Nothing to do - empty resultsPublisherConfig will cause the
+                // results to be published to S3.
+                System.out.println("Results will be published to S3 bucket " + getInstanceProperties().get(QUERY_RESULTS_BUCKET));
+            } else if ("q".equals(type)) {
+                resultsPublisherConfig.put(ResultsOutput.DESTINATION, SQSResultsOutput.SQS);
+                System.out.println("Results will be published to SQS queue " + getInstanceProperties().get(QUERY_RESULTS_QUEUE_URL));
+            } else {
+                continue;
+            }
+            break;
         }
         super.runQueries(tableProperties);
     }
@@ -143,7 +141,7 @@ public class QueryLambdaClient extends QueryCommandLineClient {
                 StsClient stsClient = StsClient.create()) {
             String accountName = stsClient.getCallerIdentity().account();
             InstanceProperties instanceProperties = S3InstanceProperties.loadGivenAccountAndInstanceId(s3Client, accountName, instanceId);
-            QueryLambdaClient queryLambdaClient = new QueryLambdaClient(s3Client, dynamoClient, sqsClient, instanceProperties);
+            QueryLambdaClient queryLambdaClient = new QueryLambdaClient(instanceProperties, s3Client, dynamoClient, sqsClient, ConsoleInput.stdIn());
             queryLambdaClient.run();
         }
     }

--- a/java/clients/src/main/java/sleeper/clients/query/QueryLambdaClient.java
+++ b/java/clients/src/main/java/sleeper/clients/query/QueryLambdaClient.java
@@ -95,7 +95,7 @@ public class QueryLambdaClient extends QueryCommandLineClient {
             out.println("Finished query processing with final state of: " + state);
         } catch (QueryTrackerException | InterruptedException e) {
             out.println("Failed to get status");
-            e.printStackTrace();
+            out.printStackTrace(e);
         }
     }
 

--- a/java/clients/src/main/java/sleeper/clients/util/console/ConsoleOutput.java
+++ b/java/clients/src/main/java/sleeper/clients/util/console/ConsoleOutput.java
@@ -74,6 +74,15 @@ public class ConsoleOutput {
     }
 
     /**
+     * Prints a stack trace to the console.
+     *
+     * @param e the error or exception
+     */
+    public void printStackTrace(Throwable e) {
+        e.printStackTrace(out);
+    }
+
+    /**
      * Creates a print writer to write to the console. Sets autoFlush to true. Uses {@link StandardCharsets#UTF_8}.
      *
      * @return the print writer

--- a/java/clients/src/test/java/sleeper/clients/admin/testutils/RunAdminClient.java
+++ b/java/clients/src/test/java/sleeper/clients/admin/testutils/RunAdminClient.java
@@ -17,6 +17,7 @@
 package sleeper.clients.admin.testutils;
 
 import sleeper.clients.admin.AdminClientTrackerFactory;
+import sleeper.clients.testutil.ExceptionWithOutput;
 import sleeper.clients.testutil.TestConsoleInput;
 import sleeper.clients.testutil.ToStringConsoleOutput;
 import sleeper.common.task.QueueMessageCount;
@@ -119,7 +120,7 @@ public class RunAdminClient {
         try {
             harness.startClient(trackers, queueClient);
         } catch (Exception e) {
-            throw new AdminClientFailedException(out.toString(), e);
+            throw new ExceptionWithOutput(out, e);
         }
         return out.toString();
     }

--- a/java/clients/src/test/java/sleeper/clients/query/QueryClientTestConstants.java
+++ b/java/clients/src/test/java/sleeper/clients/query/QueryClientTestConstants.java
@@ -31,4 +31,6 @@ public class QueryClientTestConstants {
     public static final String PROMPT_MAX_INCLUSIVE = "Is the maximum inclusive? (y/n) \n";
     public static final String PROMPT_MIN_ROW_KEY_LONG_TYPE = "Enter a minimum key for row key field key of type = LongType{} - hit return for no minimum: \n";
     public static final String PROMPT_MAX_ROW_KEY_LONG_TYPE = "Enter a maximum key for row key field key of type = LongType{} - hit return for no maximum: \n";
+    public static final String SEND_TO_S3_OPTION = "s";
+    public static final String SEND_TO_SQS_OPTION = "q";
 }

--- a/java/clients/src/test/java/sleeper/clients/query/QueryLambdaClientTest.java
+++ b/java/clients/src/test/java/sleeper/clients/query/QueryLambdaClientTest.java
@@ -1,0 +1,90 @@
+/*
+ * Copyright 2022-2026 Crown Copyright
+ *
+ * Licensed under the Apache License, Version 2.0 (the "License");
+ * you may not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ *
+ *     http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+package sleeper.clients.query;
+
+import sleeper.clients.testutil.TestConsoleInput;
+import sleeper.clients.testutil.ToStringConsoleOutput;
+import sleeper.core.iterator.closeable.CloseableIterator;
+import sleeper.core.properties.instance.InstanceProperties;
+import sleeper.core.properties.table.TableProperties;
+import sleeper.core.properties.table.TablePropertiesProvider;
+import sleeper.core.properties.table.TablePropertiesStore;
+import sleeper.core.properties.testutils.InMemoryTableProperties;
+import sleeper.core.row.Row;
+import sleeper.core.row.testutils.InMemoryRowStore;
+import sleeper.core.statestore.StateStoreProvider;
+import sleeper.core.statestore.testutils.InMemoryTransactionLogStateStore;
+import sleeper.core.statestore.testutils.InMemoryTransactionLogsPerTable;
+import sleeper.core.table.InMemoryTableIndex;
+import sleeper.core.table.TableIndex;
+import sleeper.core.util.ObjectFactory;
+import sleeper.query.core.model.Query;
+import sleeper.query.core.model.QueryException;
+import sleeper.query.core.rowretrieval.InMemoryLeafPartitionRowRetriever;
+import sleeper.query.core.rowretrieval.LeafPartitionQueryExecutor;
+import sleeper.query.core.rowretrieval.QueryExecutor;
+import sleeper.query.core.rowretrieval.QueryPlanner;
+import sleeper.query.core.tracker.InMemoryQueryTracker;
+
+import java.io.IOException;
+import java.io.UncheckedIOException;
+import java.time.Instant;
+import java.util.ArrayList;
+import java.util.List;
+
+import static sleeper.core.properties.testutils.InstancePropertiesTestHelper.createTestInstanceProperties;
+
+public class QueryLambdaClientTest {
+    private final InstanceProperties instanceProperties = createTestInstanceProperties();
+    private final TableIndex tableIndex = new InMemoryTableIndex();
+    private final ToStringConsoleOutput out = new ToStringConsoleOutput();
+    private final TestConsoleInput in = new TestConsoleInput(out.consoleOut());
+    private final InMemoryTransactionLogsPerTable transactionLogs = new InMemoryTransactionLogsPerTable();
+    private final StateStoreProvider stateStoreProvider = InMemoryTransactionLogStateStore.createProvider(instanceProperties, transactionLogs);
+    private final TablePropertiesStore tablePropertiesStore = InMemoryTableProperties.getStore(tableIndex);
+    private final InMemoryRowStore rowStore = new InMemoryRowStore();
+    private final List<Row> outputRows = new ArrayList<>();
+
+    private void runQueryClient() throws Exception {
+        new QueryLambdaClient(instanceProperties, tableIndex, tablePropertiesProvider(),
+                this::runQuery, new InMemoryQueryTracker(instanceProperties, Instant::now),
+                in.consoleIn(), out.consoleOut())
+                .run();
+    }
+
+    private TablePropertiesProvider tablePropertiesProvider() {
+        return new TablePropertiesProvider(instanceProperties, tablePropertiesStore);
+    }
+
+    private void runQuery(Query query) {
+        TableProperties tableProperties = tablePropertiesStore.loadByName(query.getTableName());
+        try (CloseableIterator<Row> rows = queryExecutor(tableProperties).execute(query)) {
+            rows.forEachRemaining(outputRows::add);
+        } catch (QueryException e) {
+            throw new RuntimeException(e);
+        } catch (IOException e) {
+            throw new UncheckedIOException(e);
+        }
+    }
+
+    private QueryExecutor queryExecutor(TableProperties tableProperties) {
+        return new QueryExecutor(
+                QueryPlanner.initialiseNow(tableProperties, stateStoreProvider.getStateStore(tableProperties)),
+                new LeafPartitionQueryExecutor(ObjectFactory.noUserJars(), tableProperties,
+                        new InMemoryLeafPartitionRowRetriever(rowStore)));
+    }
+
+}

--- a/java/clients/src/test/java/sleeper/clients/query/QueryLambdaClientTest.java
+++ b/java/clients/src/test/java/sleeper/clients/query/QueryLambdaClientTest.java
@@ -15,6 +15,9 @@
  */
 package sleeper.clients.query;
 
+import org.junit.jupiter.api.Test;
+
+import sleeper.clients.testutil.ExceptionWithOutput;
 import sleeper.clients.testutil.TestConsoleInput;
 import sleeper.clients.testutil.ToStringConsoleOutput;
 import sleeper.core.iterator.closeable.CloseableIterator;
@@ -25,6 +28,10 @@ import sleeper.core.properties.table.TablePropertiesStore;
 import sleeper.core.properties.testutils.InMemoryTableProperties;
 import sleeper.core.row.Row;
 import sleeper.core.row.testutils.InMemoryRowStore;
+import sleeper.core.schema.Schema;
+import sleeper.core.statestore.FileReference;
+import sleeper.core.statestore.FileReferenceFactory;
+import sleeper.core.statestore.StateStore;
 import sleeper.core.statestore.StateStoreProvider;
 import sleeper.core.statestore.testutils.InMemoryTransactionLogStateStore;
 import sleeper.core.statestore.testutils.InMemoryTransactionLogsPerTable;
@@ -33,6 +40,8 @@ import sleeper.core.table.TableIndex;
 import sleeper.core.util.ObjectFactory;
 import sleeper.query.core.model.Query;
 import sleeper.query.core.model.QueryException;
+import sleeper.query.core.output.ResultsOutputInfo;
+import sleeper.query.core.output.ResultsOutputLocation;
 import sleeper.query.core.rowretrieval.InMemoryLeafPartitionRowRetriever;
 import sleeper.query.core.rowretrieval.LeafPartitionQueryExecutor;
 import sleeper.query.core.rowretrieval.QueryExecutor;
@@ -41,11 +50,29 @@ import sleeper.query.core.tracker.InMemoryQueryTracker;
 
 import java.io.IOException;
 import java.io.UncheckedIOException;
-import java.time.Instant;
 import java.util.ArrayList;
 import java.util.List;
+import java.util.Map;
+import java.util.UUID;
+import java.util.stream.Collectors;
+import java.util.stream.LongStream;
 
+import static org.assertj.core.api.Assertions.assertThat;
+import static sleeper.clients.query.QueryClientTestConstants.EXIT_OPTION;
+import static sleeper.clients.query.QueryClientTestConstants.NO_OPTION;
+import static sleeper.clients.query.QueryClientTestConstants.PROMPT_MAX_INCLUSIVE;
+import static sleeper.clients.query.QueryClientTestConstants.PROMPT_MAX_ROW_KEY_LONG_TYPE;
+import static sleeper.clients.query.QueryClientTestConstants.PROMPT_MIN_INCLUSIVE;
+import static sleeper.clients.query.QueryClientTestConstants.PROMPT_MIN_ROW_KEY_LONG_TYPE;
+import static sleeper.clients.query.QueryClientTestConstants.PROMPT_QUERY_TYPE;
+import static sleeper.clients.query.QueryClientTestConstants.RANGE_QUERY_OPTION;
+import static sleeper.clients.query.QueryClientTestConstants.SEND_TO_S3_OPTION;
+import static sleeper.clients.query.QueryClientTestConstants.YES_OPTION;
+import static sleeper.core.properties.table.TableProperty.TABLE_NAME;
 import static sleeper.core.properties.testutils.InstancePropertiesTestHelper.createTestInstanceProperties;
+import static sleeper.core.properties.testutils.TablePropertiesTestHelper.createTestTableProperties;
+import static sleeper.core.schema.SchemaTestHelper.createSchemaWithKey;
+import static sleeper.core.statestore.testutils.StateStoreUpdatesWrapper.update;
 
 public class QueryLambdaClientTest {
     private final InstanceProperties instanceProperties = createTestInstanceProperties();
@@ -56,13 +83,72 @@ public class QueryLambdaClientTest {
     private final StateStoreProvider stateStoreProvider = InMemoryTransactionLogStateStore.createProvider(instanceProperties, transactionLogs);
     private final TablePropertiesStore tablePropertiesStore = InMemoryTableProperties.getStore(tableIndex);
     private final InMemoryRowStore rowStore = new InMemoryRowStore();
+    private final InMemoryQueryTracker queryTracker = new InMemoryQueryTracker(instanceProperties);
     private final List<Row> outputRows = new ArrayList<>();
 
+    @Test
+    void shouldRunRangeQuery() throws Exception {
+        // Given
+        Schema schema = createSchemaWithKey("key");
+        TableProperties tableProperties = createTable("test-table", schema);
+        List<Row> rows = LongStream.rangeClosed(0, 10)
+                .mapToObj(num -> new Row(Map.of("key", num)))
+                .collect(Collectors.toList());
+        ingestData(tableProperties, rows);
+
+        // When
+        in.enterNextPrompts(
+                SEND_TO_S3_OPTION,
+                RANGE_QUERY_OPTION,
+                NO_OPTION, YES_OPTION,
+                "3", "6",
+                EXIT_OPTION);
+        runQueryClient();
+
+        // Then
+        assertThat(out.toString())
+                .startsWith("Querying table test-table")
+                .contains(PROMPT_QUERY_TYPE +
+                        PROMPT_MIN_INCLUSIVE +
+                        PROMPT_MAX_INCLUSIVE +
+                        PROMPT_MIN_ROW_KEY_LONG_TYPE +
+                        PROMPT_MAX_ROW_KEY_LONG_TYPE +
+                        "Submitting query with id")
+                .contains("Polling query tracker\n" +
+                        "Finished query processing with final state of: COMPLETED");
+    }
+
+    private TableProperties createTable(String tableName, Schema schema) {
+        TableProperties tableProperties = createTestTableProperties(instanceProperties, schema);
+        tableProperties.set(TABLE_NAME, tableName);
+        tablePropertiesStore.save(tableProperties);
+        update(stateStore(tableProperties)).initialise(tableProperties);
+        return tableProperties;
+    }
+
+    private void ingestData(TableProperties tableProperties, List<Row> rows) throws Exception {
+        FileReference file = fileFactory(tableProperties).rootFile(UUID.randomUUID().toString(), rows.size());
+        rowStore.addFile(file.getFilename(), rows);
+        update(stateStore(tableProperties)).addFile(file);
+    }
+
+    private FileReferenceFactory fileFactory(TableProperties tableProperties) {
+        return FileReferenceFactory.from(instanceProperties, tableProperties, stateStore(tableProperties));
+    }
+
+    private StateStore stateStore(TableProperties tableProperties) {
+        return stateStoreProvider.getStateStore(tableProperties);
+    }
+
     private void runQueryClient() throws Exception {
-        new QueryLambdaClient(instanceProperties, tableIndex, tablePropertiesProvider(),
-                this::runQuery, new InMemoryQueryTracker(instanceProperties, Instant::now),
-                in.consoleIn(), out.consoleOut())
-                .run();
+        try {
+            new QueryLambdaClient(instanceProperties, tableIndex, tablePropertiesProvider(),
+                    this::runQuery, queryTracker,
+                    in.consoleIn(), out.consoleOut())
+                    .run();
+        } catch (Exception e) {
+            throw new ExceptionWithOutput(out, e);
+        }
     }
 
     private TablePropertiesProvider tablePropertiesProvider() {
@@ -71,8 +157,12 @@ public class QueryLambdaClientTest {
 
     private void runQuery(Query query) {
         TableProperties tableProperties = tablePropertiesStore.loadByName(query.getTableName());
-        try (CloseableIterator<Row> rows = queryExecutor(tableProperties).execute(query)) {
-            rows.forEachRemaining(outputRows::add);
+        try (CloseableIterator<Row> rowsIt = queryExecutor(tableProperties).execute(query)) {
+            List<Row> rows = new ArrayList<>();
+            rowsIt.forEachRemaining(rows::add);
+            outputRows.addAll(rows);
+            queryTracker.queryCompleted(query, new ResultsOutputInfo(rows.size(),
+                    List.of(new ResultsOutputLocation("s3", "test.parquet"))));
         } catch (QueryException e) {
             throw new RuntimeException(e);
         } catch (IOException e) {

--- a/java/clients/src/test/java/sleeper/clients/testutil/ExceptionWithOutput.java
+++ b/java/clients/src/test/java/sleeper/clients/testutil/ExceptionWithOutput.java
@@ -13,19 +13,19 @@
  * See the License for the specific language governing permissions and
  * limitations under the License.
  */
-package sleeper.clients.admin.testutils;
+package sleeper.clients.testutil;
 
-public class AdminClientFailedException extends RuntimeException {
+public class ExceptionWithOutput extends RuntimeException {
 
-    private final String output;
+    private final ToStringConsoleOutput output;
 
-    public AdminClientFailedException(String output, Throwable cause) {
-        super("Admin client failed with output: " + output, cause);
+    public ExceptionWithOutput(ToStringConsoleOutput output, Throwable cause) {
+        super("Failed with output: " + output, cause);
         this.output = output;
     }
 
     public String getOutput() {
-        return output;
+        return output.toString();
     }
 
 }

--- a/java/clients/src/test/java/sleeper/clients/testutil/TestConsoleInput.java
+++ b/java/clients/src/test/java/sleeper/clients/testutil/TestConsoleInput.java
@@ -39,7 +39,7 @@ public class TestConsoleInput {
             String prompt = invocation.getArgument(0);
             out.println(prompt);
             if (nextPrompts.isEmpty()) {
-                throw new IllegalStateException("User was prompted when all specified prompts were exhausted");
+                throw new IllegalStateException("User was prompted when all specified prompts were exhausted, prompt was: " + prompt);
             }
             return nextPrompts.poll();
         });

--- a/java/query/query-core/src/main/java/sleeper/query/core/tracker/TrackedQuery.java
+++ b/java/query/query-core/src/main/java/sleeper/query/core/tracker/TrackedQuery.java
@@ -15,6 +15,8 @@
  */
 package sleeper.query.core.tracker;
 
+import sleeper.query.core.output.ResultsOutputInfo;
+
 import java.time.Instant;
 import java.util.Objects;
 
@@ -48,6 +50,12 @@ public class TrackedQuery {
 
     public static Builder builder() {
         return new Builder();
+    }
+
+    public Builder toBuilder() {
+        return builder().queryId(queryId).subQueryId(subQueryId)
+                .lastUpdateTime(lastUpdateTime).expiryDate(expiryDate)
+                .lastKnownState(lastKnownState).rowCount(rowCount).errorMessage(errorMessage);
     }
 
     public String getQueryId() {
@@ -224,6 +232,27 @@ public class TrackedQuery {
         public Builder errorMessage(String errorMessage) {
             this.errorMessage = errorMessage;
             return this;
+        }
+
+        /**
+         * Provides information on the results of the query.
+         *
+         * @param  outputInfo the information
+         * @return            the builder
+         */
+        public Builder outputInfo(ResultsOutputInfo outputInfo) {
+            return rowCount(outputInfo.getRowCount())
+                    .error(outputInfo.getError());
+        }
+
+        /**
+         * Provides the error.
+         *
+         * @param  error the error
+         * @return       the builder
+         */
+        public Builder error(Exception error) {
+            return errorMessage(error == null ? null : error.getMessage());
         }
 
         public TrackedQuery build() {

--- a/java/query/query-core/src/test/java/sleeper/query/core/tracker/InMemoryQueryTracker.java
+++ b/java/query/query-core/src/test/java/sleeper/query/core/tracker/InMemoryQueryTracker.java
@@ -43,6 +43,10 @@ public class InMemoryQueryTracker implements QueryStatusReportListener, QueryTra
     private final Map<String, TrackedQuery> queryIdToStatus = new HashMap<>();
     private final Map<String, TrackedQuery> subQueryIdToStatus = new HashMap<>();
 
+    public InMemoryQueryTracker(InstanceProperties instanceProperties) {
+        this(instanceProperties, Instant::now);
+    }
+
     public InMemoryQueryTracker(InstanceProperties instanceProperties, Supplier<Instant> timeSupplier) {
         this.instanceProperties = instanceProperties;
         this.timeSupplier = timeSupplier;

--- a/java/query/query-core/src/test/java/sleeper/query/core/tracker/InMemoryQueryTracker.java
+++ b/java/query/query-core/src/test/java/sleeper/query/core/tracker/InMemoryQueryTracker.java
@@ -1,0 +1,219 @@
+/*
+ * Copyright 2022-2026 Crown Copyright
+ *
+ * Licensed under the Apache License, Version 2.0 (the "License");
+ * you may not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ *
+ *     http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+package sleeper.query.core.tracker;
+
+import sleeper.core.properties.instance.InstanceProperties;
+import sleeper.query.core.model.LeafPartitionQuery;
+import sleeper.query.core.model.Query;
+import sleeper.query.core.output.ResultsOutputInfo;
+
+import java.time.Duration;
+import java.time.Instant;
+import java.util.HashMap;
+import java.util.List;
+import java.util.Map;
+import java.util.Objects;
+import java.util.Optional;
+import java.util.function.Function;
+import java.util.function.Supplier;
+import java.util.stream.Stream;
+
+import static sleeper.core.properties.instance.QueryProperty.QUERY_TRACKER_ITEM_TTL_IN_DAYS;
+
+/**
+ * An in-memory implementation of the query tracker. Tracks and reports on the status of queries.
+ */
+public class InMemoryQueryTracker implements QueryStatusReportListener, QueryTrackerStore {
+
+    private final InstanceProperties instanceProperties;
+    private final Supplier<Instant> timeSupplier;
+    private final Map<String, TrackedQuery> queryIdToStatus = new HashMap<>();
+    private final Map<String, TrackedQuery> subQueryIdToStatus = new HashMap<>();
+
+    public InMemoryQueryTracker(InstanceProperties instanceProperties, Supplier<Instant> timeSupplier) {
+        this.instanceProperties = instanceProperties;
+        this.timeSupplier = timeSupplier;
+    }
+
+    @Override
+    public TrackedQuery getStatus(String queryId) throws QueryTrackerException {
+        TrackedQuery status = queryIdToStatus.get(queryId);
+        if (status == null) {
+            throw new RuntimeException("Query ID was not found: " + queryId);
+        }
+        return status;
+    }
+
+    @Override
+    public TrackedQuery getStatus(String queryId, String subQueryId) throws QueryTrackerException {
+        TrackedQuery status = subQueryIdToStatus.get(subQueryId);
+        if (status == null) {
+            throw new RuntimeException("Sub-query ID was not found: " + subQueryId);
+        }
+        if (!Objects.equals(status.getQueryId(), queryId)) {
+            throw new RuntimeException("Sub-query ID was found but query ID was not found: " + queryId);
+        }
+        return status;
+    }
+
+    @Override
+    public List<TrackedQuery> getAllQueries() {
+        return streamAllQueries().toList();
+    }
+
+    /**
+     * Streams through the status of all queries and sub-queries.
+     *
+     * @return the query statuses
+     */
+    public Stream<TrackedQuery> streamAllQueries() {
+        return Stream.concat(queryIdToStatus.values().stream(), subQueryIdToStatus.values().stream());
+    }
+
+    @Override
+    public List<TrackedQuery> getQueriesWithState(QueryState state) {
+        return streamAllQueries().filter(query -> Objects.equals(state, query.getLastKnownState())).toList();
+    }
+
+    @Override
+    public List<TrackedQuery> getFailedQueries() {
+        return streamAllQueries()
+                .filter(query -> query.getLastKnownState() == QueryState.FAILED
+                        || query.getLastKnownState() == QueryState.PARTIALLY_FAILED)
+                .toList();
+    }
+
+    @Override
+    public void queryQueued(Query query) {
+        queryIdToStatus.put(query.getQueryId(), newQuery(query, QueryState.QUEUED));
+    }
+
+    @Override
+    public void queryInProgress(Query query) {
+        upsertQuery(query,
+                () -> newQuery(query, QueryState.IN_PROGRESS),
+                status -> updateQuery(status).lastKnownState(QueryState.IN_PROGRESS).build());
+    }
+
+    @Override
+    public void queryInProgress(LeafPartitionQuery leafQuery) {
+        upsertQuery(leafQuery,
+                () -> newQuery(leafQuery, QueryState.IN_PROGRESS),
+                status -> updateQuery(status).lastKnownState(QueryState.IN_PROGRESS).build());
+    }
+
+    @Override
+    public void subQueriesCreated(Query query, List<LeafPartitionQuery> subQueries) {
+        for (LeafPartitionQuery subQuery : subQueries) {
+            subQueryIdToStatus.put(subQuery.getSubQueryId(), newQuery(subQuery, QueryState.QUEUED));
+        }
+    }
+
+    @Override
+    public void queryCompleted(Query query, ResultsOutputInfo outputInfo) {
+        upsertQuery(query,
+                () -> newQueryBuilder(query, QueryState.COMPLETED).outputInfo(outputInfo).build(),
+                status -> updateQuery(status).lastKnownState(QueryState.COMPLETED).outputInfo(outputInfo).build());
+    }
+
+    @Override
+    public void queryCompleted(LeafPartitionQuery leafQuery, ResultsOutputInfo outputInfo) {
+        upsertQuery(leafQuery,
+                () -> newQueryBuilder(leafQuery, QueryState.COMPLETED).outputInfo(outputInfo).build(),
+                status -> updateQuery(status).lastKnownState(QueryState.COMPLETED).outputInfo(outputInfo).build());
+    }
+
+    @Override
+    public void queryFailed(Query query, Exception e) {
+        upsertQuery(query,
+                () -> newQueryBuilder(query, QueryState.FAILED).error(e).build(),
+                status -> updateQuery(status).lastKnownState(QueryState.FAILED).error(e).build());
+    }
+
+    @Override
+    public void queryFailed(String queryId, Exception e) {
+        upsertQuery(queryId,
+                () -> newQueryBuilder(queryId, QueryState.FAILED).error(e).build(),
+                status -> updateQuery(status).lastKnownState(QueryState.FAILED).error(e).build());
+    }
+
+    @Override
+    public void queryFailed(LeafPartitionQuery leafQuery, Exception e) {
+        upsertQuery(leafQuery,
+                () -> newQueryBuilder(leafQuery, QueryState.FAILED).error(e).build(),
+                status -> updateQuery(status).lastKnownState(QueryState.FAILED).error(e).build());
+    }
+
+    private void upsertQuery(Query query, Supplier<TrackedQuery> createNew, Function<TrackedQuery, TrackedQuery> update) {
+        upsertQuery(query.getQueryId(), createNew, update);
+    }
+
+    private void upsertQuery(String queryId, Supplier<TrackedQuery> createNew, Function<TrackedQuery, TrackedQuery> update) {
+        queryIdToStatus.put(queryId,
+                Optional.ofNullable(queryIdToStatus.get(queryId))
+                        .map(update)
+                        .orElseGet(createNew));
+    }
+
+    private void upsertQuery(LeafPartitionQuery query, Supplier<TrackedQuery> createNew, Function<TrackedQuery, TrackedQuery> update) {
+        subQueryIdToStatus.put(query.getSubQueryId(),
+                Optional.ofNullable(subQueryIdToStatus.get(query.getSubQueryId()))
+                        .map(update)
+                        .orElseGet(createNew));
+    }
+
+    private TrackedQuery newQuery(Query query, QueryState state) {
+        return newQueryBuilder(query, state).build();
+    }
+
+    private TrackedQuery newQuery(LeafPartitionQuery leafQuery, QueryState state) {
+        return newQueryBuilder(leafQuery, state).build();
+    }
+
+    private TrackedQuery.Builder newQueryBuilder(Query query, QueryState state) {
+        return newQueryBuilder(query.getQueryId(), state);
+    }
+
+    private TrackedQuery.Builder newQueryBuilder(LeafPartitionQuery leafQuery, QueryState state) {
+        return newQueryBuilder(leafQuery.getQueryId(), leafQuery.getSubQueryId(), state);
+    }
+
+    private TrackedQuery.Builder newQueryBuilder(String queryId, QueryState state) {
+        return newQueryBuilder(queryId, null, state);
+    }
+
+    private TrackedQuery.Builder newQueryBuilder(String queryId, String subQueryId, QueryState state) {
+        Instant now = timeSupplier.get();
+        return TrackedQuery.builder()
+                .queryId(queryId)
+                .subQueryId(subQueryId)
+                .lastUpdateTime(now)
+                .expiryDate(now.plus(queryTtl()))
+                .lastKnownState(state);
+    }
+
+    private TrackedQuery.Builder updateQuery(TrackedQuery status) {
+        Instant now = timeSupplier.get();
+        return status.toBuilder()
+                .lastUpdateTime(now)
+                .expiryDate(now.plus(queryTtl()));
+    }
+
+    private Duration queryTtl() {
+        return Duration.ofDays(instanceProperties.getLong(QUERY_TRACKER_ITEM_TTL_IN_DAYS));
+    }
+
+}

--- a/java/system-test/system-test-data-generation/src/main/java/sleeper/systemtest/datageneration/MultipleQueries.java
+++ b/java/system-test/system-test-data-generation/src/main/java/sleeper/systemtest/datageneration/MultipleQueries.java
@@ -25,6 +25,7 @@ import software.amazon.awssdk.services.sqs.model.ReceiveMessageResponse;
 import software.amazon.awssdk.services.sts.StsClient;
 
 import sleeper.clients.query.QueryLambdaClient;
+import sleeper.clients.util.console.ConsoleInput;
 import sleeper.configuration.properties.S3TableProperties;
 import sleeper.core.key.Key;
 import sleeper.core.properties.table.TablePropertiesProvider;
@@ -79,7 +80,7 @@ public class MultipleQueries {
 
     public void run() {
         TablePropertiesProvider tablePropertiesProvider = S3TableProperties.createProvider(systemTestProperties, s3Client, dynamoClient);
-        QueryLambdaClient queryLambdaClient = new QueryLambdaClient(s3Client, dynamoClient, sqsClient, systemTestProperties);
+        QueryLambdaClient queryLambdaClient = new QueryLambdaClient(systemTestProperties, s3Client, dynamoClient, sqsClient, ConsoleInput.stdIn());
 
         Schema schema = tablePropertiesProvider.getByName(tableName).getSchema();
         RangeFactory rangeFactory = new RangeFactory(schema);

--- a/java/system-test/system-test-data-generation/src/main/java/sleeper/systemtest/datageneration/MultipleQueries.java
+++ b/java/system-test/system-test-data-generation/src/main/java/sleeper/systemtest/datageneration/MultipleQueries.java
@@ -26,6 +26,7 @@ import software.amazon.awssdk.services.sts.StsClient;
 
 import sleeper.clients.query.QueryLambdaClient;
 import sleeper.clients.util.console.ConsoleInput;
+import sleeper.clients.util.console.ConsoleOutput;
 import sleeper.configuration.properties.S3TableProperties;
 import sleeper.core.key.Key;
 import sleeper.core.properties.table.TablePropertiesProvider;
@@ -80,7 +81,7 @@ public class MultipleQueries {
 
     public void run() {
         TablePropertiesProvider tablePropertiesProvider = S3TableProperties.createProvider(systemTestProperties, s3Client, dynamoClient);
-        QueryLambdaClient queryLambdaClient = new QueryLambdaClient(systemTestProperties, s3Client, dynamoClient, sqsClient, ConsoleInput.stdIn());
+        QueryLambdaClient queryLambdaClient = new QueryLambdaClient(systemTestProperties, s3Client, dynamoClient, sqsClient, ConsoleOutput.stdOut(), ConsoleInput.stdIn());
 
         Schema schema = tablePropertiesProvider.getByName(tableName).getSchema();
         RangeFactory rangeFactory = new RangeFactory(schema);

--- a/java/system-test/system-test-data-generation/src/main/java/sleeper/systemtest/datageneration/MultipleQueries.java
+++ b/java/system-test/system-test-data-generation/src/main/java/sleeper/systemtest/datageneration/MultipleQueries.java
@@ -24,10 +24,12 @@ import software.amazon.awssdk.services.sqs.model.Message;
 import software.amazon.awssdk.services.sqs.model.ReceiveMessageResponse;
 import software.amazon.awssdk.services.sts.StsClient;
 
+import sleeper.clients.api.QuerySender;
 import sleeper.clients.query.QueryLambdaClient;
 import sleeper.clients.util.console.ConsoleInput;
 import sleeper.clients.util.console.ConsoleOutput;
 import sleeper.configuration.properties.S3TableProperties;
+import sleeper.configuration.table.index.DynamoDBTableIndex;
 import sleeper.core.key.Key;
 import sleeper.core.properties.table.TablePropertiesProvider;
 import sleeper.core.range.Range;
@@ -37,8 +39,10 @@ import sleeper.core.row.ResultsBatch;
 import sleeper.core.row.Row;
 import sleeper.core.row.serialiser.JSONResultsBatchSerialiser;
 import sleeper.core.schema.Schema;
+import sleeper.core.table.TableIndex;
 import sleeper.core.util.LoggedDuration;
 import sleeper.query.core.model.Query;
+import sleeper.query.runner.tracker.DynamoDBQueryTracker;
 import sleeper.systemtest.configuration.SystemTestProperties;
 import sleeper.systemtest.configuration.SystemTestRandomDataSettings;
 
@@ -80,8 +84,13 @@ public class MultipleQueries {
     }
 
     public void run() {
+        TableIndex tableIndex = new DynamoDBTableIndex(systemTestProperties, dynamoClient);
         TablePropertiesProvider tablePropertiesProvider = S3TableProperties.createProvider(systemTestProperties, s3Client, dynamoClient);
-        QueryLambdaClient queryLambdaClient = new QueryLambdaClient(systemTestProperties, s3Client, dynamoClient, sqsClient, ConsoleOutput.stdOut(), ConsoleInput.stdIn());
+        QueryLambdaClient queryLambdaClient = new QueryLambdaClient(
+                systemTestProperties, tableIndex, tablePropertiesProvider,
+                QuerySender.toSqs(systemTestProperties, tablePropertiesProvider, sqsClient),
+                new DynamoDBQueryTracker(systemTestProperties, dynamoClient),
+                ConsoleInput.stdIn(), ConsoleOutput.stdOut());
 
         Schema schema = tablePropertiesProvider.getByName(tableName).getSchema();
         RangeFactory rangeFactory = new RangeFactory(schema);

